leader/undirected_ring/sync_higham_przytycka/sync_higham_przytycka.go

```diff
@@ -0,0 +1,413 @@
+	uhp{}.send(Left, node, &state, mailbox[Left])
+	uhp{}.send(Right, node, &state, mailbox[Right])
+
+	// log.Printf("[%d] State after round: %#v", node.GetIndex(), state)
+
+	return false
+}
+
+// endregion
+
+// region uhp::Processor (scope: unidirectional)
+// region uhp::Processor::test
+func (proc uhpProcessor) testLeader(curr uhpEnvelope) bool {
+	// envelope returned to owner
+	return curr == proc.Last
+}
+
+func (proc uhpProcessor) testReceive() bool {
+	switch proc.Status {
+	case Active:
+		return true
+	case TempLeader:
+		return true
+	default:
+		return false
+	}
+}
+
+func (proc uhpProcessor) testFinished() bool {
+	switch proc.Status {
+	case Active:
+		return false
+	case TempLeader:
+		return false
+	case TrueLeader:
+		return true
+	case Dead:
+		return true
+	default:
+		return true
+	}
+}
+
+// endregion
+
+// region uhp::Processor::process
+func (proc *uhpProcessor) process(received *uhpMessage) (optSend *uhpMessage) {
+	switch {
+	case received == nil:
+		return nil
+	case received.ENVELOPE != nil:
+		return proc.processEnvelope(*received.ENVELOPE)
+	case received.BROADCAST != nil:
+		// log.Printf(LOG_PREFIX+"[BROADCAST:%d] received. processor: %#v", &proc.Id, &proc.Dir, *received.BROADCAST, proc)
+		return proc.processBroadcast(*received.BROADCAST)
+	case proc.Status == Active:
+		return &uhpMessage{}
+	default:
+		return nil
+	}
+}
+
+func (proc *uhpProcessor) processEnvelope(curr uhpEnvelope) *uhpMessage {
+	if curr.Max < proc.Id {
+		curr.Max = proc.Id
+	}
+	// log.Printf(LOG_PREFIX+"Received envelope %#v", proc.Id, proc.Dir, curr)
+	if proc.testLeader(curr) {
+		proc.Status = TempLeader
+		log.Printf(LOG_PREFIX+"UHP found one of two leaders: %#v", proc.Id, proc.Dir, proc)
+		return &uhpMessage{BROADCAST: &curr.Max}
+	}
+
+	// log.Printf(LOG_PREFIX+"comparing... [%#v] vs [%#v]", proc.Id, proc.Dir, proc.Last, curr)
+	if drop, promote := proc.Promoter.Test(proc.Last, curr); !drop {
+		if promote {
+			log.Printf(LOG_PREFIX+"envelope promoted!", proc.Id, proc.Dir)
+			proc.Promoter.Promote(&curr)
+		}
+		proc.Last = curr
+		return &uhpMessage{ENVELOPE: &curr}
+	} else {
+		log.Printf(LOG_PREFIX+"dropping envelope!", proc.Id, proc.Dir)
+		return &uhpMessage{}
+	}
+}
+
+func (proc *uhpProcessor) processBroadcast(curr uhpBroadcast) (toSend *uhpMessage) {
+	switch proc.Status {
+	case Active:
+		toSend = &uhpMessage{BROADCAST: &curr}
+	case TempLeader:
+		log.Printf(LOG_PREFIX+"[BROADCAST:%d] performed round trip. demoting UHP leader...", proc.Id, proc.Dir, curr)
+		toSend = nil
+	default:
+		log.Panicf(LOG_PREFIX+"[BROADCAST:%d] invalid state received broadcast", proc.Id, proc.Dir, curr)
+	}
+
+	if proc.Id == curr {
+		proc.Status = TrueLeader
+		log.Printf(LOG_PREFIX+"[BROADCAST:%d] processor is true leader", proc.Id, proc.Dir, curr)
+	} else {
+		proc.Status = Dead
+	}
+	return toSend
+}
+
+// endregion
+
+// endregion
+
+// region driver
+func run(v lib.Node) {
+	v.StartProcessing()
+	uhp{}.initiateElection(v)
```

Comment:
Please refactor `uhp{}.initializeElection()` as `initialize()` (you can create uhp inside the function)